File: code/ex496.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n1 = nums1.length, n2 = nums2.length;
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = n2 - 1; i >= 0; --i) {
            for (int j = i + 1; j < n2; ++j) {
                if (nums2[j] > nums2[i]) {
                    map.put(nums2[i], nums2[j]);
                    break;
                }
            }
        }
        
        int[] res = new int[n1];
        for (int i = 0; i < n1; i++)
            res[i] = map.getOrDefault(nums1[i], -1);
        
        return res;
    }
}

File: code/ex496.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> greater;
        stack<int> stk;    // Monotonously decreasing
        for (int i = nums2.size() - 1; i >= 0; --i) {
            while (!stk.empty() && nums2[i] >= stk.top())
                stk.pop();
            greater[nums2[i]] = stk.empty() ? -1 : stk.top();
            stk.push(nums2[i]);
        }
        
        // Get results
        vector<int> res;
        for (int num : nums1)
            res.push_back(greater[num]);
        
        return res;
    }
};